tests/pytests/unit/states/test_ldap.py

```diff
@@ -0,0 +1,387 @@
+"""Test cases for the ``ldap`` state module
+
+This code is gross.  I started out trying to remove some of the
+duplicate code in the test cases, and before I knew it the test code
+was an ugly second implementation.
+
+I'm leaving it for now, but this should really be gutted and replaced
+with something sensible.
+"""
+import copy
+
+import pytest
+import salt.states.ldap
+from salt.utils.oset import OrderedSet
+from salt.utils.stringutils import to_bytes
+
+# emulates the LDAP database.  each key is the DN of an entry and it
+# maps to a dict which maps attribute names to sets of values.
+db = {}
```

Comment:
we want to avoid using a mutable object on the global/module level. Is this possible?

Reply:
Thought I had fixed that one, will take a look.

Reply:
Make it a fixture.